manual/grammar/grammar.ebnf

```diff
@@ -0,0 +1,188 @@
+extras  ::=  { multi_comment single_comment /\s|\\\r?\n|\u00A0/ }
+
+conflicts  ::=  { { pub_var } { fresh_var } { msg_var_or_nullary_fun } { temporal_var } { nat_var } { pub_var fresh_var msg_var_or_nullary_fun temporal_var nat_var } { pub_var fresh_var msg_var_or_nullary_fun nat_var } { accountability_lemma diff_lemma } { lemma diff_lemma } { nary_app predicate_ref } { nullary_fun nary_app msg_var_or_nullary_fun } }
+
+externals  ::=  { multi_comment single_comment }
+
+precedences  ::=  { { 'NESTED' 'FUNCTION' 'VARIABLE' 'EXPONENTIAL' 'MULTIPLY' 'ADD' 'MUL_SET' 'TUPLE' 'NULLARY_FUN' 'ATOM' 'LOGICAL_NOT' 'EXCLUSIVE_OR' 'LOGICAL_AND' 'LOGICAL_OR' 'LOGICAL_IMPLICATION' 'LOGICAL_IFF' 'CHAIN_GOAL' } { 'NON_DIFF' 'DIFF' } { 'REPLICATION' 'EVENT' 'CHOICE' 'PROCESS_LET' 'LOOKUP' 'CONDITIONAL' } }
+
+word  ::=  ident
+
+rules:
+  theory  ::=  'theory' (ident: theory_name) 'begin' _body_item* 'end' /./*
+  _body_item  ::=  preprocessor | _signature_spec | global_heuristic | tactic | process | let | export | _rule | restriction | case_test | _lemma | formal_comment
+  preprocessor  ::=  ifdef | define | include
+  ifdef  ::=  '#ifdef' _ifdef_formula _body_item* ('#else' _body_item*)? '#endif'
+  define  ::=  '#define' ident
+  include  ::=  '#include' '"' (param -> path) '"'
+  _ifdef_formula  ::=  ifdef_nested | ifdef_or | ifdef_and | ifdef_not | ident
+  ifdef_nested  ::=  NESTED('(' _ifdef_formula ')')
+  ifdef_or  ::=  <LOGICAL_OR(_ifdef_formula '|' _ifdef_formula)
+  ifdef_and  ::=  <LOGICAL_AND(_ifdef_formula '&' _ifdef_formula)
+  ifdef_not  ::=  LOGICAL_NOT('not' _ifdef_formula)
+  _signature_spec  ::=  built_ins | functions | equations | predicates | macros | options
+  built_ins  ::=  'builtins' ':' built_in (',' built_in)* ','?
+  built_in  ::=  'diffie-hellman' | 'hashing' | 'symmetric-encryption' | 'asymmetric-encryption' | 'signing' | 'bilinear-pairing' | 'xor' | 'multiset' | 'natural-numbers' | 'revealing-signing' | 'locations-report' | 'reliable-channel' | 'dest-pairing' | 'dest-signing' | 'dest-symmetric-encryption' | 'dest-asymmetric-encryption'
+  functions  ::=  >('functions' ':' _function_sym (',' _function_sym)* ','?)
+  _function_sym  ::=  function_pub | function_private | function_destructor | function_custom
+  function_pub  ::=  (ident: function_identifier) '/' (natural: arity)
+  function_private  ::=  (ident: function_identifier) '/' (natural: arity) '[' 'private' ']'
+  function_destructor  ::=  (ident: function_identifier) '/' (natural: arity) '[' 'destructor' ']'
```

Comment:
Here the grammar does not permit a list of function attributes, e.g., declaring a function `[private, destructor]`, but this is permitted by Tamarin.